### PR TITLE
[bugfix] Logging only on `not should_accumulate()` during training

### LINKED ISSUE #5405: W&B logger not working as expected with accumulate_grad_batches>1
## 🐛 Bug

When logging inside training step to `wandb` logger and using `accumulate_grad_batches > 1` the behavior is not as expected. Similar issue as in #4304 for Tensorboard (which was closed and the fix was merged in #4738).

First half with `accumulate_grad_batches == 1`, second with `accumulate_grad_batches == 8`:

image

Moreover, the logging steps are `accumulate_grad_batches * number_of_backward_passes` and so when using `LearningRateMonitor`, the logger refuses to log anything (similar to #4811) saying:

```
wandb: WARNING Step must only increase in log calls.  Step 499 < 2497; dropping {'lr-AdamW': 4.9900000000000005e-06}.
```

### To Reproduce

Sorry for not using the BoringModel, updated the example from #4304: 

```python
import os

import torch
from torch.nn import functional as F
from torch.utils.data import DataLoader, random_split

import pytorch_lightning as pl
from pytorch_lightning.loggers import WandbLogger
from torchvision.datasets.mnist import MNIST
from torchvision import transforms

class LitClassifier(pl.LightningModule):
    def __init__(self, hidden_dim=128, learning_rate=1e-3):
        super().__init__()
        self.save_hyperparameters()

        self.l1 = torch.nn.Linear(28 * 28, self.hparams.hidden_dim)
        self.l2 = torch.nn.Linear(self.hparams.hidden_dim, 10)

    def forward(self, x):
        x = x.view(x.size(0), -1)
        x = torch.relu(self.l1(x))
        x = torch.relu(self.l2(x))
        return x

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)
        self.log("train_loss",loss)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.cross_entropy(y_hat, y)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=self.hparams.learning_rate)

def run_test(accumulate_grad_batches, batch_size, num_workers=4):
    dataset = MNIST('', train=True, download=True, transform=transforms.ToTensor())
    mnist_train, mnist_val = random_split(dataset, [55000, 5000])
    train_loader = DataLoader(mnist_train,batch_size)
    val_loader = DataLoader(mnist_val,batch_size)

    model = LitClassifier()

    trainer = pl.Trainer(
        logger=WandbLogger(name="bug", project='.....', save_dir=".", log_model=False),
        accumulate_grad_batches=accumulate_grad_batches,
        max_epochs=2
        )
    trainer.fit(model, train_loader, val_loader)

run_test(1,32)
run_test(8,32)
```

### Expected behavior

Take a mean (or whatever) of the values logged at the same step and not at every forward pass.

### Environment

- pytorch-lightning: 1.1.2
- PyTorch: 1.7.1
- OS: Linux
- How you installed PyTorch: pip
- Python version: 3.7.0
- wandb: 0.10.12

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR fixes visual logging with accumulated_grad_batches > 1.

Solution: We can't assume the metrics can be averaged, so we will log only when `optimizer_step` will be called.

Previously

Screenshot 2021-01-08 at 13 16 19

Now.

Screenshot 2021-01-08 at 12 55 57

Code used to generate the visualization

```
def test_logging_with_accumulate_grad_batches(tmpdir):
    class LitClassifier(pl.LightningModule):
        def __init__(self, hidden_dim=128, learning_rate=1e-3):
            super().__init__()
            self.save_hyperparameters()

            self.l1 = torch.nn.Linear(28 * 28, self.hparams.hidden_dim)
            self.l2 = torch.nn.Linear(self.hparams.hidden_dim, 10)

            self.train_acc = pl.metrics.Accuracy()

        def forward(self, x):
            x = x.view(x.size(0), -1)
            x = torch.relu(self.l1(x))
            return self.l2(x)

        def training_step(self, batch, batch_idx):
            x, y = batch
            y_hat = self(x)
            loss = F.cross_entropy(y_hat, y)
            self.log('train_acc', self.train_acc(y_hat, y), on_step=True, on_epoch=True, prog_bar=True)
            self.log("train_loss",loss)
            return loss

        def validation_step(self, batch, batch_idx):
            x, y = batch
            y_hat = self(x)
            loss = F.cross_entropy(y_hat, y)

        def configure_optimizers(self):
            return torch.optim.Adam(self.parameters(), lr=self.hparams.learning_rate)

    def run_test(model, max_epochs, accumulate_grad_batches, batch_size, num_workers=4):
        dataset = MNIST('', train=True, download=True, transform=transforms.ToTensor())
        mnist_train, mnist_val = random_split(dataset, [55000, 5000])
        train_loader = DataLoader(mnist_train,batch_size)
        val_loader = DataLoader(mnist_val,batch_size)

        trainer = pl.Trainer(
            logger=WandbLogger(name="bug", project='.....', save_dir=".", log_model=False),
            accumulate_grad_batches=accumulate_grad_batches,
            limit_train_batches=100,
            log_every_n_steps=5,
            max_epochs=max_epochs
            )
        trainer.fit(model, train_loader, val_loader)

    model = LitClassifier()

    run_test(model, 3, 1, 32)
    run_test(model, 3, 8, 32)
```

Fixes #5405 <- this [links related issue to this PR](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/loggers/test_tensorboard.py [hunk 1]
```diff
@@ -213,19 +213,22 @@ def test_tensorboard_with_accummulated_gradients(mock_log_metrics, expected, tmp
...
         def training_step(self, batch, batch_idx):
             output = self.layer(batch)
             loss = self.loss(batch, output)
             self.log('count', self._count, on_step=True, on_epoch=True)
             self.log('loss', loss, on_step=True, on_epoch=True)
 
-            if self.trainer.logger_connector.should_update_logs:
-                self._indexes.append(self._count)
+            if not self.trainer.train_loop.should_accumulate():
+                if self.trainer.logger_connector.should_update_logs or self.trainer.fast_dev_run is True:
```
Comment (carmocca) [inline tests/loggers/test_tensorboard.py:229]:
```suggestion
                if self.trainer.logger_connector.should_update_logs:
```

I guess you added this while debugging

**Comment by carmocca** [comment]
Not familiar with W&B. Can you see the raw data for the epoch graph? Doesn't look right, does it?

**Comment by tchaton** [comment]
> Not familiar with W&B. Can you see the raw data for the epoch graph? Doesn't look right, does it?

It looks good to me ! I am creating a new Trainer and Trainer from 3 epochs.

**Review by carmocca — APPROVED** [review]

**Review by teddykoker — APPROVED** [review]
Great fix :)

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -158,7 +158,7 @@ def cache_training_step_metrics(self, opt_closure_result):
         self.logged_metrics.update(logged_metrics_tmp)
         self.cached_results.legacy_batch_log_metrics.update(logged_metrics_tmp)
 
-    def log_metrics(self, metrics, grad_norm_dic, step=None, log_train_step_metrics=False):
+    def log_metrics(self, metrics, grad_norm_dic, step=None):
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:161]: is it used by a user, right? then let's hold back compatibility with API and add a warning...

**Review by Borda — APPROVED** [review]
lgtm, just check back-compatibity and chlog

**Review by ananyahjha93 — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:161]: It is internal. LoggerConnector class

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:161]: Not it is not.
